cdc/owner/changefeed.go

```diff
@@ -561,36 +569,60 @@ func (c *changefeed) asyncExecDDL(ctx cdcContext.Context, job *timodel.Job) (don
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-		ddlEvent.Query, err = addSpecialComment(ddlEvent.Query)
-		if err != nil {
-			log.Error("add special comment fail", zap.String("changefeed", c.id),
-				zap.String("Query", ddlEvent.Query), zap.Error(err))
-			return false, errors.Trace(err)
-		}
 
-		c.ddlEventCache = ddlEvent
-		if c.redoManager.Enabled() {
-			err = c.redoManager.EmitDDLEvent(ctx, ddlEvent)
-			if err != nil {
-				return false, err
+		c.ddlEventCache = ddlEvents
+		c.ddlEventDone = make([]bool, len(ddlEvents))
+		for _, ddlEvent := range ddlEvents {
+			if c.redoManager.Enabled() {
+				err = c.redoManager.EmitDDLEvent(ctx, ddlEvent)
+				if err != nil {
+					return false, err
+				}
 			}
 		}
 	}
-	if job.BinlogInfo.TableInfo != nil && c.schema.IsIneligibleTableID(job.BinlogInfo.TableInfo.ID) {
-		log.Warn("ignore the DDL job of ineligible table",
-			zap.String("changefeed", c.id), zap.Reflect("job", job))
-		return true, nil
-	}
-	done, err = c.sink.emitDDLEvent(ctx, c.ddlEventCache)
-	if err != nil {
-		return false, err
+
+	jobDone := true
+	for i, ddlEvent := range c.ddlEventCache {
+		if c.ddlEventDone[i] == true {
+			continue
+		}
+		eventDone, err := c.asyncExecDDLEvent(ctx, ddlEvent)
+		if err != nil {
+			return false, err
+		}
+		if eventDone == true {
+			c.ddlEventDone[i] = true
+		}
+		jobDone = jobDone && eventDone
```

Comment:
if some eventDones = false ,not all , is there any problem in the line 512 ?

Reply:
> if some eventDones = false ,not all , is there any problem in the line 512 ?

If some events are not done, `asyncExecDDLJob()` will return false. In next run, the events not done yet will be  executed by `asyncExecDDLEvent()`. In other word, `asyncExecDDLJob()` will return true until all of the events are completed.